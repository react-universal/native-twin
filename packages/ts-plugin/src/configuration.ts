import { TwindPluginConfiguration } from './types';

export class ConfigurationManager {
  private static readonly defaultConfiguration: TwindPluginConfiguration = {
    tags: ['tw', 'apply', 'css', 'styled', 'variants'],
    attributes: ['tw', 'class', 'className', 'variants'],
    styles: ['style', 'styled'],
    debug: false,
    enable: true,
  };

  private readonly _configUpdatedListeners = new Set<() => void>();

  public get config(): TwindPluginConfiguration {
    return this._configuration;
  }

  private _configuration: TwindPluginConfiguration = {
    ...ConfigurationManager.defaultConfiguration,
    tags: [...ConfigurationManager.defaultConfiguration.tags],
  };

  public updateFromPluginConfig(config: Partial<TwindPluginConfiguration> = {}): void {
    const mergedConfig = {
      ...ConfigurationManager.defaultConfiguration,
      ...config,
    };

    this._configuration = {
      ...mergedConfig,
      debug: 'true' == String(mergedConfig.debug),
      enable: 'false' != String(mergedConfig.enable),
    };

    for (const listener of this._configUpdatedListeners) {
      listener();
    }
  }

  public onUpdatedConfig(listener: () => void): void {
    this._configUpdatedListeners.add(listener);
  }
}
