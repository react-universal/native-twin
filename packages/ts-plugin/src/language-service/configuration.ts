export interface TwindPluginConfiguration {
  readonly configFile?: string;
  readonly tags: ReadonlyArray<string>;
  readonly attributes: ReadonlyArray<string>;
  readonly styles: ReadonlyArray<string>;
  readonly debug?: boolean;
  readonly enable: boolean;
}

export class ConfigurationManager {
  static readonly pluginName = 'ts-styled-plugin-tw';
  static readonly VARIANT_MARKER_RULE = /^â€¦$/;
  private static readonly defaultConfiguration: TwindPluginConfiguration = {
    tags: ['tw', 'apply', 'css', 'styled', 'variants'],
    attributes: ['tw', 'class', 'className', 'variants'],
    styles: ['style', 'styled'],
    debug: false,
    enable: true,
  };

  private readonly _configUpdatedListeners = new Set<() => void>();

  public get config(): TwindPluginConfiguration {
    return this._configuration;
  }

  private _configuration: TwindPluginConfiguration = {
    ...ConfigurationManager.defaultConfiguration,
    tags: [...ConfigurationManager.defaultConfiguration.tags],
  };

  public updateFromPluginConfig(config: Partial<TwindPluginConfiguration> = {}): void {
    const mergedConfig = {
      ...ConfigurationManager.defaultConfiguration,
      ...config,
    };

    this._configuration = {
      ...mergedConfig,
      debug: 'true' == String(mergedConfig.debug),
      enable: 'false' != String(mergedConfig.enable),
    };

    for (const listener of this._configUpdatedListeners) {
      listener();
    }
  }

  public onUpdatedConfig(listener: () => void): void {
    this._configUpdatedListeners.add(listener);
  }
}
