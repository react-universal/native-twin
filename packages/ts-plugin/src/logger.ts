import type { Logger } from 'typescript-template-language-service-decorator';
import ts from 'typescript/lib/tsserverlibrary';
import { pluginName } from './constants/config.constants';

export class LanguageServiceLogger implements Logger {
  private readonly info: ts.server.PluginCreateInfo;

  constructor(info: ts.server.PluginCreateInfo) {
    this.info = info;
  }

  log(message: string): void {
    this.info.project.projectService.logger.info(`[${pluginName}] ${message}`);
  }
}
