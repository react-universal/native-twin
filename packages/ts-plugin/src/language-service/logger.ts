import type { Logger } from 'typescript-template-language-service-decorator';
import ts from 'typescript/lib/tsserverlibrary';
import { ConfigurationManager } from './configuration';

export class LanguageServiceLogger implements Logger {
  private readonly info: ts.server.PluginCreateInfo;

  constructor(info: ts.server.PluginCreateInfo) {
    this.info = info;
  }

  log(message: string): void {
    this.info.project.projectService.logger.info(
      `[${ConfigurationManager.pluginName}]: ${message}`,
    );
  }
}
