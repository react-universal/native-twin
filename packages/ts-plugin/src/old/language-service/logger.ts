import type { Logger } from 'typescript-template-language-service-decorator';
import ts from 'typescript/lib/tsserverlibrary';
import { ConfigurationManager } from './configuration';

export class LanguageServiceLogger implements Logger {
  private readonly pluginInfo: ts.server.PluginCreateInfo;

  constructor(info: ts.server.PluginCreateInfo) {
    this.pluginInfo = info;
  }

  private sendMessage(message: string, kind: ts.server.Msg) {
    this.pluginInfo.project.projectService.logger.msg(
      `[${ConfigurationManager.pluginName}]: ${message}`,
      kind,
    );
  }

  log(message: string): void {
    this.sendMessage(message, ts.server.Msg.Info);
  }
  perf(message: string): void {
    this.sendMessage(message, ts.server.Msg.Perf);
  }
}
