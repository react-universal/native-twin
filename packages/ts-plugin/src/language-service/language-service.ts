import {
  TemplateContext,
  TemplateLanguageService,
} from 'typescript-template-language-service-decorator';
import ts from 'typescript/lib/tsserverlibrary';
import * as vscode from 'vscode-languageserver-types';
import { CompletionToken } from '../types';
import { NativeTailwindService } from './createIntellisense';
import { LanguageServiceLogger } from './logger';
import { ParsedCompletionRule, parse } from './parser';

export class NativeTailwindLanguageService implements TemplateLanguageService {
  constructor(
    private _logger: LanguageServiceLogger,
    private _intellisense: NativeTailwindService,
  ) {}
  getCompletionsAtPosition(
    context: TemplateContext,
    position: ts.LineAndCharacter,
  ): ts.WithMetadata<ts.CompletionInfo> {
    const text = context.text;
    const textOffset = context.toOffset(position);
    const parsedRule = parse(text, textOffset, true);
    return translateCompletionItemsToCompletionInfo(context, {
      isIncomplete: true,
      items: Array.from([
        ...this._intellisense.completions.classes.values(),
        ...this._intellisense.completions.variants.values(),
      ]).map((completion, index) => getCompletionItem(context, completion, parsedRule, index)),
    });
  }
}

function getCompletionItem(
  context: TemplateContext,
  completion: CompletionToken,
  rule: ParsedCompletionRule,
  sortedIndex: number,
): vscode.CompletionItem {
  return {
    kind: getCompletionTokenKind(completion),
    data: completion.kind,
    label:
      rule.prefix && completion.name !== '&' && completion.kind == 'class'
        ? completion.name.slice(rule.prefix.length + 1)
        : completion.name,
    preselect: false,
    filterText: rule.name,
    sortText: sortedIndex.toString().padStart(8, '0'),
    textEdit: {
      newText:
        rule.prefix && completion.name !== '&' && completion.kind == 'class'
          ? completion.name.slice(rule.prefix.length + 1)
          : completion.name,
      range: {
        start: context.toPosition(rule.loc.start),
        end: context.toPosition(rule.loc.end),
      },
    },
  };
}

export function getCompletionTokenKind(completion: CompletionToken) {
  if (completion.kind == 'variant') {
    return vscode.CompletionItemKind.Module;
  }
  if (completion.isColor) return vscode.CompletionItemKind.Color;
  return vscode.CompletionItemKind.Property;
}

// function findCompletion(rule: ParsedCompletionRule, intellisense: IntellisenseService) {
//   for (const completion of intellisense.classes.values()) {
//     const label =
//       rule.prefix &&
//       completion.kind == 'class' &&
//       completion.name.startsWith(rule.prefix + '-')
//         ? completion.name.slice(rule.prefix.length + 1)
//         : completion.name;
//   }
// }

function translateCompletionItemsToCompletionInfo(
  context: TemplateContext,
  items: vscode.CompletionList,
): ts.WithMetadata<ts.CompletionInfo> {
  return {
    metadata: {
      isIncomplete: items.isIncomplete,
    },
    isGlobalCompletion: false,
    isMemberCompletion: false,
    isNewIdentifierLocation: false,
    entries: items.items.map((x) => translateCompletionEntry(context, x)),
  };
}

function translateCompletionEntry(
  context: TemplateContext,
  item: vscode.CompletionItem,
): ts.CompletionEntry {
  let replacementSpan: ts.TextSpan | undefined;
  if ('range' in item.textEdit!) {
    replacementSpan = {
      start: context.toOffset(item.textEdit.range.start),
      length:
        context.toOffset(item.textEdit.range.end) -
        context.toOffset(item.textEdit.range.start),
    };
  }
  return {
    name: item.label,
    kind: item.kind
      ? translateCompletionItemKind(context.typescript, item.kind)
      : context.typescript.ScriptElementKind.unknown,
    kindModifiers: getKindModifiers(item),
    sortText: item.sortText || item.label,
    insertText: item.textEdit && item.textEdit.newText,
    replacementSpan,
  };
}

function getKindModifiers(item: vscode.CompletionItem): string {
  if (item.kind === vscode.CompletionItemKind.Color) {
    return 'color';
  }

  return '';
}

function translateCompletionItemKind(
  typescript: typeof ts,
  kind: vscode.CompletionItemKind,
): ts.ScriptElementKind {
  switch (kind) {
    case vscode.CompletionItemKind.Module:
      return typescript.ScriptElementKind.moduleElement;
    case vscode.CompletionItemKind.Property:
      return typescript.ScriptElementKind.memberVariableElement;
    case vscode.CompletionItemKind.Unit:
    case vscode.CompletionItemKind.Value:
      return typescript.ScriptElementKind.constElement;
    case vscode.CompletionItemKind.Variable:
      return typescript.ScriptElementKind.variableElement;
    case vscode.CompletionItemKind.Enum:
      return typescript.ScriptElementKind.enumElement;
    case vscode.CompletionItemKind.EnumMember:
      return typescript.ScriptElementKind.enumMemberElement;
    case vscode.CompletionItemKind.Keyword:
      return typescript.ScriptElementKind.keyword;
    case vscode.CompletionItemKind.Constant:
      return typescript.ScriptElementKind.constElement;
    case vscode.CompletionItemKind.Color:
      return typescript.ScriptElementKind.primitiveType;
    case vscode.CompletionItemKind.Reference:
      return typescript.ScriptElementKind.alias;
    case vscode.CompletionItemKind.Snippet:
    case vscode.CompletionItemKind.Text:
      return typescript.ScriptElementKind.string;
    default:
      return typescript.ScriptElementKind.unknown;
  }
}
