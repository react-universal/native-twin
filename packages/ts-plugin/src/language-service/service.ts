import {
  TemplateContext,
  TemplateLanguageService,
} from 'typescript-template-language-service-decorator';
import StandardScriptSourceHelper from 'typescript-template-language-service-decorator/lib/standard-script-source-helper';
import ts from 'typescript/lib/tsserverlibrary';
import { CreateIntellisenseFn } from '../intellisense/createIntellisense';
import { createCompletionEntries, createCompletionEntryDetails } from '../utils';
import { ConfigurationManager } from './configuration';
import { LanguageServiceLogger } from './logger';
import { StandardTemplateSourceHelper } from './source-helper';

export class TailwindLanguageService implements TemplateLanguageService {
  configManager: ConfigurationManager;
  logger: LanguageServiceLogger;
  templateSourceHelper: StandardTemplateSourceHelper;
  intellisense: CreateIntellisenseFn;
  constructor(
    typescript: typeof ts,
    info: ts.server.PluginCreateInfo,
    configManager: ConfigurationManager,
    intellisense: CreateIntellisenseFn,
  ) {
    this.intellisense = intellisense;
    this.configManager = configManager;
    this.logger = new LanguageServiceLogger(info);
    this.templateSourceHelper = new StandardTemplateSourceHelper(
      typescript,
      configManager,
      new StandardScriptSourceHelper(typescript, info.project),
    );
  }

  getCompletionsAtPosition(
    templateContext: TemplateContext,
    position: ts.LineAndCharacter,
  ): ts.WithMetadata<ts.CompletionInfo> {
    let originalList = Array.from(this.intellisense.classes.values());
    const templateClasses = new Set(templateContext.text.split(/\s+/).filter(Boolean));
    const isEmptyCompletion = templateContext.text.charAt(position.character - 1) == ' ';
    const prevText = templateContext.text.slice(0, position.character);

    if (isEmptyCompletion) {
      // Filter insertedClasses
      originalList = originalList.filter((i) => !templateClasses.has(i.name));
    }
    if (prevText == '-') {
      originalList = originalList
        // .filter((x) => x.)
        .map((x) => {
          const className = `-${x.name}`;
          const completion = this.intellisense.classes.get(className);
          return {
            ...x,
            className,
            css: '',
            sheet: completion,
          };
        });
    }
    if (prevText.length > 0 && !isEmptyCompletion) {
      const prevClasses = prevText.split(/\s+/).filter(Boolean);
      const completion = prevClasses[prevClasses.length - 1]!;
      originalList = originalList.filter(
        (i) => !templateClasses.has(i.name) && i.name.includes(completion),
      );
    }
    return createCompletionEntries(originalList);
  }

  getCompletionEntryDetails(
    context: TemplateContext,
    position: ts.LineAndCharacter,
    name: string,
  ): ts.CompletionEntryDetails {
    const utility = this.intellisense.classes.get(name)!;
    return createCompletionEntryDetails(utility);
  }
}
