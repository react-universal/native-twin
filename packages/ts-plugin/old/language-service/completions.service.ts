import { TinyColor } from '@ctrl/tinycolor';
import { TemplateContext } from 'typescript-template-language-service-decorator';
import ts from 'typescript/lib/tsserverlibrary';
import * as vscode from 'vscode-languageserver-types';
import { RuntimeTW } from '@native-twin/core';
import { ClassCompletionToken, CompletionToken } from '../types';
import { debugLog } from '../utils';
import { NativeTailwindIntellisense } from './intellisense.service';
import { ParsedCompletionRule, parse } from './parser';

export function getCompletionsAtPosition(
  context: TemplateContext,
  position: ts.LineAndCharacter,
  intellisense: NativeTailwindIntellisense,
): ts.WithMetadata<ts.CompletionInfo> {
  const text = context.text;
  const textOffset = context.toOffset(position);
  const parsedRule = parse(text, textOffset, true);

  // ts.SyntaxKind.NoSubstitutionTemplateLiteral
  // ts.SyntaxKind.TemplateExpression
  return {
    isGlobalCompletion: false,
    isMemberCompletion: false,
    isNewIdentifierLocation: false,
    entries: getCompletionEntries(context, parsedRule, intellisense),
  };
}

function getCompletionEntries(
  context: TemplateContext,
  rule: ParsedCompletionRule,
  intellisense: NativeTailwindIntellisense,
): ts.CompletionEntry[] {
  const result: ts.CompletionEntry[] = [];
  const { completions } = intellisense;
  const hasScreenValue = rule.variants.some((x) => completions().screens.has(x.value));
  const screens = hasScreenValue ? [] : Array.from(completions().variants.values());
  let index = screens.length;
  if (!hasScreenValue && rule.raw == '') {
    screens.forEach((x, i) => {
      result.push({
        kind: ts.ScriptElementKind.alias,
        name: x.name,
        sortText: i.toString().padStart(8, '0'),
        kindModifiers: getCompletionTokenKind(x),
        labelDetails: { description: 'Screen Media Query', detail: 'Screens' },
      });
    });
  }
  for (const token of completions().classes.values()) {
    const nextToken = createCompletionEntry(context, rule, token, index);
    if (!nextToken) continue;
    result.push(nextToken);
    index++;
  }
  return result;
}

function createCompletionEntry(
  context: TemplateContext,
  rule: ParsedCompletionRule,
  completion: CompletionToken,
  sortedIndex: number,
): ts.CompletionEntry | null {
  let name = completion.name;
  let insertText: string | undefined;
  let replacementSpan: ts.TextSpan | undefined = {
    start: context.toOffset(context.toPosition(rule.loc.start)),
    length:
      context.toOffset(context.toPosition(rule.loc.end)) -
      context.toOffset(context.toPosition(rule.loc.start)),
  };
  if (rule.prefix != '') {
    if (name.startsWith(rule.name)) {
      name = completion.name.slice(rule.prefix.length);
      replacementSpan = undefined;
      if (name.startsWith('-')) {
        name = name.slice(1);
      }
    } else {
      return null;
    }
  }
  if (rule.prefix == '') {
    if ((rule.name != '' || rule.raw != '') && !completion.name.startsWith(rule.name)) {
      return null;
    }
    replacementSpan = undefined;
    // insertText = completion.name.slice(rule.raw.length);
    name = completion.name;
  }
  return {
    kind: getCompletionTokenKind(completion),
    kindModifiers: getKindModifiers(completion),
    name,
    sourceDisplay: [
      {
        kind: 'color',
        text:
          completion.kind == 'class' && completion.themeValue
            ? `${completion.property}: ${completion.themeValue}`
            : '',
      },
    ],
    sortText: sortedIndex.toString().padStart(8, '0'),
    insertText,
    replacementSpan,
  };
}

export function getCompletionTokenKind(completion: CompletionToken): ts.ScriptElementKind {
  if (completion.kind == 'variant') {
    return ts.ScriptElementKind.moduleElement;
  }
  if (completion.isColor) return ts.ScriptElementKind.primitiveType;
  return ts.ScriptElementKind.string;
}

function getKindModifiers(item: CompletionToken): string {
  if (item.kind == 'variant') return '';
  if (item.isColor) {
    return 'color';
  }

  return '';
}

export function getDocumentation(data: ClassCompletionToken, tw: RuntimeTW) {
  if (!data.property || !data.themeValue) return '';
  const result: string[] = [];
  // result.push('***Css Rules*** \n\n');
  // result.push(`${'```css\n'}${data.css}${'\n```'}`);
  // result.push('\n\n');
  const resolver = tw(data.name);
  debugLog(
    {
      symbol: 'ðŸ“—',
      msg: `____DATA_____`,
      value: resolver,
    },
    true,
  );
  const prop = resolver.reduce(
    (prev, current) => {
      current.declarations.forEach((x) => {
        prev[x.prop] = x.value;
      });
      return {
        ...prev,
      };
    },
    {} as Record<string, any>,
  );
  result.push('***React Native StyleSheet*** \n\n');
  result.push(`${'```json\n'}${JSON.stringify(prop, null, 2)}${'\n```'}`);
  return result.join('\n\n');
}

export function getCompletionEntryDetailsDisplayParts(
  suggestion: ClassCompletionToken,
): ts.SymbolDisplayPart[] {
  if (
    suggestion.isColor &&
    suggestion.themeValue &&
    (suggestion.themeValue.startsWith('#') || suggestion.themeValue.startsWith('rgb'))
  ) {
    const hex = new TinyColor(suggestion.themeValue);
    if (hex.isValid) {
      return [
        {
          kind: 'color',
          text: hex.toHexString(),
        },
      ];
    }
  }
  return [];
}

export function createCompletionEntryDetails(
  item: ClassCompletionToken,
  twRuntime: RuntimeTW,
): ts.CompletionEntryDetails {
  const displayParts = getCompletionEntryDetailsDisplayParts(item);
  return {
    name: item.name,
    kind: ts.ScriptElementKind.string,
    kindModifiers: displayParts.length > 0 ? 'color' : '',
    displayParts,
    documentation: [
      {
        kind: vscode.MarkupKind.Markdown,
        text: getDocumentation(item, twRuntime),
      },
    ],
  };
}
