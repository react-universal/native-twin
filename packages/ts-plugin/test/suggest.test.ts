import { ConfigurationManager } from '../src/language-service/configuration';
import { NativeTailwindIntellisense } from '../src/language-service/intellisense.service';
import { LanguageServiceLogger } from '../src/language-service/logger';

const config = new ConfigurationManager();
// @ts-expect-error
const logger: LanguageServiceLogger = {
  log: () => {},
};
const intellisense = new NativeTailwindIntellisense(logger, config);
describe('TS PLUGIN', () => {
  it('Complete suggestion list', () => {
    expect(intellisense.completions().classes.size).toStrictEqual(5530);
  });
  it('Enumerate completions', () => {
    expect(Array.from(intellisense.completions().classes, ([name]) => name)).toMatchSnapshot(
      'Completions Snapshot',
    );
  });
});
