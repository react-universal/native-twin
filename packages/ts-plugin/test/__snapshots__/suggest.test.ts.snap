// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`suggest when additional modifiers exist 1`] = `
[
  "bg-red",
  "bg-red",
  "bg-red",
  "bg-red",
  "bg-red-50",
  "bg-red-100",
  "bg-red-200",
  "bg-red-300",
  "bg-red-400",
  "bg-red-500",
  "bg-red-600",
  "bg-red-700",
  "bg-red-800",
  "bg-red-900",
  "bg-red-50",
  "bg-red-100",
  "bg-red-200",
  "bg-red-300",
  "bg-red-400",
  "bg-red-500",
  "bg-red-600",
  "bg-red-700",
  "bg-red-800",
  "bg-red-900",
  "bg-red-50",
  "bg-red-100",
  "bg-red-200",
  "bg-red-300",
  "bg-red-400",
  "bg-red-500",
  "bg-red-600",
  "bg-red-700",
  "bg-red-800",
  "bg-red-900",
  "bg-red-50",
  "bg-red-100",
  "bg-red-200",
  "bg-red-300",
  "bg-red-400",
  "bg-red-500",
  "bg-red-600",
  "bg-red-700",
  "bg-red-800",
  "bg-red-900",
  "bg-repeat-round",
]
`;
